2018-10-10 In-class notes - Data in Social Context

**What does statistical thinking look like and how do I interpret statistical output in Python?**

*The goal of today's class is to better understand the difference between simplistic data-driven thinking and statistical thinking. In a nutshell, the type of statistics I mean here are meant to help answer the questions "Is it real?" or "Does it matter?"*

*Suppose the top two students in a class have a 98.1% semester grade and a 98.2% semester grade respectively. Which one is the better student? Simplistic data-driven thinking says student 1 is better because they have a higher grade. But any number of factors could have produced the difference, from the timing of assignments for other classes to the relative weighting of skills associated with each student's strengths. Few people would argue that Student 1 is conclusively a stronger student.*

*Statistics provide a mathematically-informed way to make such decisions. In today's lesson, we'll test the likely range of population means of scores, had other similar students had the same assignment, and whether students who like N Sync are actually more successful in their assignments than those that prefer Backstreet Boys.*

In [1]:
#import pandas
import pandas as pd

In [2]:
#read data in from github (no download)
indata = pd.read_excel('https://github.com/ndporter/pythonDiSC/raw/master/Foundations_grades_F18_anon.xlsx')

In [5]:
#look at what you've got
indata.head()

classNum  ID  nsLong  bsbLong  compLong  songList  noXmas  for   if  elif  \
0   class1   1     2.0      5.0       3.0       2.0     0.0  1.0  1.0   1.0   
1   class1   2     5.0      5.0       5.0       5.0     3.0  3.0  5.0   5.0   
2   class1   3     5.0      5.0       5.0       5.0     5.0  5.0  5.0   5.0   
3   class1   4     5.0      5.0       5.0       5.0     5.0  5.0  5.0   3.0   
4   class1   5     4.0      5.0       5.0       5.0     5.0  5.0  4.0   5.0   

     ...      total  minSpent  bestBand  hot  qual  prod  members  other  \
0    ...       35.0     300.0       NaN  NaN   NaN   NaN      NaN    NaN   
1    ...      101.0      60.0         B  1.0   0.0   0.0      0.0    0.0   
2    ...      101.0      60.0         N  0.0   1.0   0.0      0.0    0.0   
3    ...      103.0      30.0         B  0.0   0.0   1.0      0.0    0.0   
4    ...      102.0     120.0         B  0.0   1.0   0.0      0.0    0.0   

   ansCompProd  bsbLonger  
0         1.00        0.0  
1         4.25        1.0  
2         1.00        0.0  
3         4.25        1.0  
4         4.25        1.0  

[5 rows x 30 columns]

In [4]:
#remove missing data - but not all, only cases that don't have an overall score
valid = indata.dropna(subset=['total'])

**A good first question: what's the average score in the class?**

To find the mean, describe the relevant column (after dealing with missing data). The other lines mean the following:
- *count*: number of cases/rows included in analysis
- *mean*: mean value (sum of values divided by count)
- *std*: standard deviation (a measure of how widely dispersed values are; a low std means most people had similar scores; 21 is quite high for assignment grades
- *min*: lowest value in data - check for unusual values (like negative scores)
- *25/50/75%*: quartiles - 25% of students have scores below 77.75, 50% below 95.5, and 75% below 101
- *max*: highest value in data - check for unusual values (like scores more than the total possible = 105)

In [6]:
valid['total'].describe()

count     44.000000
mean      85.159091
std       21.341022
min       35.000000
25%       77.750000
50%       95.500000
75%      101.000000
max      105.000000
Name: total, dtype: float64

**Ok, now we know the average score we saw, but what might the average score be for the same assignment in another similar class?**

*For this question, we need to construct what's called a confidence interval, or CI. CIs give a range of values that samples (groups of actual cases) taken from the same population (potential cases) might have for something, in this case the mean overall score. The population here might be VT students who registered for any section of Data in Social Context, and the sample is students in this section who completed the assignment.*

*Most scientists use a standard of 95% CI's, that is a range that we're 95% certain the population mean falls into based on the data. This means we accept a 5% probability that the mean is outside our calculated confidence interval.

In [16]:
#We'll need some statistical tools here
#Notice that some modules have multiple layers
#If we just imported statsmodels, we'd have to type statsmodels.stats.weightstats.COMMAND every time
import statsmodels.stats.weightstats as stats

In [9]:
#The zconfint command returns the lower and higher bounds of the CI (and defaults to 95% CI).
stats.zconfint(valid['total'])

(78.853337763333158, 91.464844054848655)

To interpret the values above in words:

*we are 95% certain that the true or population mean score, were every DiSC student to have completed the assignment, would fall between 78.85 and 91.46.*

**Hey, I wonder if people who think N Sync are better than Backstreet Boys are smarter (and do better work)? Or vice versa?**

*For this question, we need not only to find the means, but to statistically compare them to see if any difference is likely to be due to chance. There will almost always be a difference between the means, even if it's very small, like a fraction of a point. So what's important is to know whether the difference is meaningful.*

*In tests like this, most scientists use a p<0.05 test statistic, meaning we are willing to accept a 5% or 1 in 20 probability of finding a statistical difference when there isn't a true difference in the population. This prevents having very many false positives and is generally a conservative way of reporting findings.*

In [17]:
#First we need to make subsets of the groups we're comparing (people who thing N Sync are better or Backstreet Boys are better)
#The first expression can be read "valid where the value of the bestBand variable in valid is the string 'N'."
nsLove=valid[valid['bestBand']=='N']
#We do the same thing for Backstreet Boys. We're ignoring the "Other" and missing answers for now.
bsbLove=valid[valid['bestBand']=='B']

In [11]:
#Scroll over to 'bestBand' and check that it worked
nsLove.head()

classNum  ID  nsLong  bsbLong  compLong  songList  noXmas  for   if  elif  \
2    class1   3     5.0      5.0       5.0       5.0     5.0  5.0  5.0   5.0   
5    class1   6     5.0      5.0       5.0       5.0     5.0  5.0  5.0   3.0   
10   class2  11     5.0      5.0       3.0       5.0     5.0  5.0  5.0   5.0   
12   class2  13     5.0      5.0       5.0       5.0     5.0  5.0  5.0   5.0   
19   class2  20     4.0      5.0       5.0       5.0     3.0  5.0  3.0   3.0   

    ...    time  bonus  total  minSpent  bestBand  hot  qual  prod  members  \
2   ...     5.0    5.0  101.0      60.0         N  0.0   1.0   0.0      0.0   
5   ...     5.0    5.0   99.0      60.0         N  0.0   0.0   1.0      0.0   
10  ...     5.0    5.0  103.0     300.0         N  0.0   1.0   0.0      0.0   
12  ...     5.0    5.0   98.0      40.0         N  0.0   1.0   0.0      1.0   
19  ...     5.0    5.0   92.0     210.0         N  0.0   0.0   0.0      1.0   

    other  
2     0.0  
5     0.0  
10    0.0  
12    0.0  
19    0.0  

[5 rows x 28 columns]

In [12]:
#Let's see the means
nsLove['total'].describe()

count     13.000000
mean      92.923077
std       15.332218
min       48.000000
25%       91.000000
50%       99.000000
75%      101.000000
max      105.000000
Name: total, dtype: float64

In [13]:
bsbLove['total'].describe()

count     25.000000
mean      86.600000
std       19.050372
min       44.000000
25%       78.000000
50%       94.000000
75%      102.000000
max      105.000000
Name: total, dtype: float64

**92.9 certainly looks larger than 86.6, but is it meaningful or the result of random differences in samples?**

*Let's use a T-test (again a procedure in statsmodels) to find out.*

In [14]:
#For reference, this is how you would calculate the difference automatically
nsLove['total'].mean()-bsbLove['total'].mean()


6.3230769230769255

In [15]:
#Test the difference of means using a T-test
import statsmodels.stats.weightstats as stats
stats.ttest_ind(nsLove['total'],bsbLove['total'])

(1.0332313152236081, 0.30838715138763295, 36.0)

To interpret the above:
- The first value is the *T-Statistic*, which is used to conduct the test. Larger values are associated with more probability of a meaningful difference
- The **second value, the p-value of the test,** is the key here. If p is less than 0.05, we say there is a statistically significant difference between the groups; e.g. they're not random samples drawn from the same population. Remember there is still a 1 in 20 probability that you're wrong, but if the test value p is decreased, we then increase the probability of not finding a difference even when there is a true difference.
- The third value is the degrees of freedom for the test, which is based on the number of cases. More cases provide more degrees of freedom, which allows for more statistical confidence even with the same differences in means. This is why you can't just ask the three people next to you which band is better and assume that more people in the US (or even the class) prefer whichever two of the three of them say.

**That's it for Wednesday. On Friday (now below), we talked about other statistical tests to work with different kinds of variables and how to decide what you need. Thanks!**



**\*Friday 2018-10-12 Variable types, how to choose a statistics, and Chi-squared tests\***

If you haven't already, run the code above to ensure you have the correct packages imported and have the *indata* and *valid* dataframes created.

The first thing we did today was take a step back from how to run statistics and asked how to decide on a statistic. To do that, we need some terminology.

An **interval variable** is a number, like your score on an assignment. Values can not only be compared (larger, smaller, equal) but can be stated in relation to each other (i.e. Student B scored 5 less than student A; Student B missed twice as many points as student A).

An **ordinal variable** still has order (larger, smaller, equal) but the units are not consistent or comparable. If I ask "how satisfied are you with your life overall right now?" and give the options ["not at all satisfied", "somewhat satisfied", "very satisfied"], not only do I not know whether the distance between "not at all" and "somewhat" is the same as from "somewhat" to "very", the distances will change from person to person. Even so, the three categories have a clear order.

A **nominal** or **categorical variable** has neither quantity nor order, but is instead made of categories or names (thus why they're called categorical or nominal). Hair color is categorical. Red is not more or less of something than blond, unless you're scientifically measuring the volume of certain pigments. Sometimes numbers will be assigned to categories of a nominal variable, but they still can't be compared in the same way as interval or ordinal variables.


Depending on the level of a variable (interval, ordinal or nominal), there are different techniques that can be used to summarize and compare it. To take a simple example, a nominal variable can't have a mean. If half of people have blond hair and half have brown, you can't just say that people typically have dirty blond hair.

We filled out part of a table in class, a simplified version of which is below. We'll ignore ordinal variables for now.

**Dependent variable** types are in the rows. The dependent variable is the variable you're really interested in knowing about. If your research question is why some students score higher than others, the dependent variable is *score* and your **independent variables** are anything you think might affect that score. 

Relationships aren't always causal; you may just want to know whether things tend to happen together. In that case, your dependent variable is whichever you're most interested in explaining.

Here is the table:
<img src='https://github.com/ndporter/pythonDiSC/raw/master/VariableTypesTable.JPG'>

The cells *in italic* are the ones we're concerned with here. The others [in brackets] we haven't covered, but I include them for completeness.


We can summarize information without statistics, but we need the statistics in this table to tell us how precise our summary is.

We can calculate the average (mean) value of a variable and summarize it in one number. *BUT* a **confidence interval of the mean** allows us to estimate what the average could have been if our sample was different (in our data, if the students in our class were different).

We can make a crosstab table of two categorical variables and show how many cases are in each cell. *BUT* a **chi-squared test** allows us to test whether any differences we see in the proportions in each cell are likely due to chance or actually meaningful.

We can estimate separate means of the same variable for two groups of people (or any type of cases). *BUT* a **T-test** allows us to test whether the difference between those means (they very rarely are exactly the same) is likely due to chance or actually meaningful.

Above, we've already calculated a confidence interval and done a t-test of the difference of means. Below, we will make a cross-tabulation table and run a chi-squared test on it. After that, I will demonstrate how to estimate OLS regression, ANOVA and logistic regression, but *you will not be responsible for understanding or using them in this class*.

In [5]:
#pandas has built-in cross-tab functionality that makes very nice tables
#the crosstab command follows the format (rows,columns)
tab = pd.crosstab(valid['classNum'],valid['bestBand'])

In [6]:
#Exampine the table we made
#To read the first row...
#In class1: 3 students think Backstreet Boys are better, 2 students think N Sync
tab

bestBand   B  N  O
classNum          
class1     3  2  0
class2    18  6  1
class3     4  5  1

In [15]:
#It looks like class2 is different from the others; they seem to really like BSB
#But is the difference maybe just random? Let's check with a chi-squared test
#Don't forget you can search for documentation or examples for any of these procedures to see more about them

#Import the chi-squared procedure from scipy (the package name is short for "scientific python")
from scipy.stats.contingency import chi2_contingency

In [8]:
#the values returned are the Chi-square statistic, p-value, degrees of freedom, and the expected counts
#p-value is most important for you - remember we're typically willing to accept a 1 in 20 (0.05) probabily a positive result is due to chance
chi2_contingency(tab)

(3.5606153846153843,
 0.46872169381747264,
 4,
 array([[  3.125,   1.625,   0.25 ],
        [ 15.625,   8.125,   1.25 ],
        [  6.25 ,   3.25 ,   0.5  ]]))

In [14]:
from statsmodels.formula.api import ols
import statsmodels.api as sm
#ignore any warning this throws (it's not an error)
#it has to do with a part of the modules we aren't using

**You're done! You can look below if you're interested in how to run an OLS regression and have enough statistical background to understand what the output means. Otherwise, you can get started on the assignment!

In [10]:
#Which classes if any do better on the assignment?
model = ols('total ~ classNum',data=valid).fit()

In [13]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  total   R-squared:                       0.028
Model:                            OLS   Adj. R-squared:                 -0.020
Method:                 Least Squares   F-statistic:                    0.5886
Date:                Fri, 12 Oct 2018   Prob (F-statistic):              0.560
Time:                        10:33:58   Log-Likelihood:                -195.97
No. Observations:                  44   AIC:                             397.9
Df Residuals:                      41   BIC:                             403.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept             90.1667      8.797     10.250      0.000      72.401     107.933
classNum[T.class2]    -7.7963      9.725     -0.802      0.427     -27.437      11.845
classNum[T.class3]    -0.8939     10.936     -0.082      0.935     -22.980      21.192
==============================================================================
Omnibus:                        8.757   Durbin-Watson:                   1.936
Prob(Omnibus):                  0.013   Jarque-Bera (JB):                8.853
Skew:                          -1.098   Prob(JB):                       0.0120
Kurtosis:                       3.059   Cond. No.                         5.97
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""